state/cleanup_test.go

```diff
@@ -43,6 +43,12 @@ func (s *CleanupSuite) SetUpTest(c *gc.C) {
 
 }
 
+func (s *CleanupSuite) TestCleanupDyingApplicationNoUnits(c *gc.C) {
+	mysql := s.AddTestingApplication(c, "mysql", s.AddTestingCharm(c, "mysql"))
+	c.Assert(mysql.Destroy(), jc.ErrorIsNil)
```

Comment:
Who is running the cleanup here? Is there a potential race?

Reply:
so this is the code that fast-paths Destroy to immediately destroy the application if there are no units and/or relations. Nothing calls cleanup. Ultimately Cleanup effectively just calls app.Destroy under that same circumstance.